x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,141 @@
+                        "keystore"));
+                return;
+            }
+            final List<X509Certificate> caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment).stream()
+                .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1).collect(Collectors.toList());
+            if (caCertificates.size() != 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain a " +
+                    "single PrivateKey entry where the associated certificate is a CA certificate"));
+
+            } else {
+                final TimeValue expiration = TimeValue.timeValueSeconds(ENROLL_API_KEY_EXPIRATION_SEC);
+                final List<RoleDescriptor> roleDescriptors = new ArrayList<>(1);
+                final String[] clusterPrivileges = { "enroll" };
+                final RoleDescriptor roleDescriptor = new RoleDescriptor("create_enrollment_token", clusterPrivileges, null, null);
+                roleDescriptors.add(roleDescriptor);
+                CreateApiKeyRequest apiRequest = new CreateApiKeyRequest("enrollment_token_API_key_" + UUIDs.base64UUID(),
+                    roleDescriptors, expiration);
+                final String fingerprint = SslUtil.calculateFingerprint(caCertificates.get(0));
+                createEnrolmentToken(fingerprint, apiRequest, listener);
+            }
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private void createEnrolmentToken(String fingerprint, CreateApiKeyRequest apiRequest,
+                                      ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            generator.generateApiKey(securityContext.getAuthentication(), apiRequest,
+                ActionListener.wrap(
+                    CreateApiKeyResponse -> {
+                        final NodeInfo nodeInfo = nodeService.info(false, false, false, false, false, false,
+                            true, false, false, false, false);
+                        final HttpInfo httpInfo = nodeInfo.getInfo(HttpInfo.class);
+                        final String address = httpInfo.getAddress().publishAddress().toString();
+
+                        final XContentBuilder builder = JsonXContent.contentBuilder();
+                        builder.startObject();
+                        builder.field("adr", address);
+                        builder.field("fgr", fingerprint);
+                        builder.field("key", CreateApiKeyResponse.getKey().toString());
+                        builder.endObject();
+                        final String jsonString = Strings.toString(builder);
+                        final String token = Base64.getEncoder().encodeToString(jsonString.getBytes(StandardCharsets.UTF_8));
+
```

Comment:
maybe log the enrollment token pre-encoding in TRACE level ?

Reply:
Not sure if we should log this material...

Reply:
@albertzaharovits what's your take on this?

Reply:
We usually do not log credentials. In this case they are short lived, but on the other hand they are powerful (add a new node). I would not log it.